File: shared/src/main/java/com/runtracker/shared/data/repository/GymRepository.kt
package com.runtracker.shared.data.repository

import com.runtracker.shared.data.db.*
import com.runtracker.shared.data.model.*
import kotlinx.coroutines.flow.Flow
import java.util.*

class GymRepository(
    private val exerciseDao: ExerciseDao,
    private val gymWorkoutDao: GymWorkoutDao,
    private val workoutTemplateDao: WorkoutTemplateDao,
    private val exerciseHistoryDao: ExerciseHistoryDao
) {
    // Exercises
    fun getAllExercises(): Flow<List<Exercise>> = exerciseDao.getAllExercisesFlow()
    
    fun getExercisesByMuscleGroup(muscleGroup: MuscleGroup): Flow<List<Exercise>> = 
        exerciseDao.getExercisesByMuscleGroupFlow(muscleGroup)
    
    fun getExercisesByEquipment(equipment: Equipment): Flow<List<Exercise>> = 
        exerciseDao.getExercisesByEquipmentFlow(equipment)
    
    fun searchExercises(query: String): Flow<List<Exercise>> = 
        exerciseDao.searchExercisesFlow(query)
    
    suspend fun getExerciseById(id: Long): Exercise? = exerciseDao.getExerciseById(id)
    
    suspend fun insertExercise(exercise: Exercise): Long = exerciseDao.insertExercise(exercise)
    
    suspend fun insertExercises(exercises: List<Exercise>) = exerciseDao.insertExercises(exercises)
    
    suspend fun initializeDefaultExercises() {
        if (exerciseDao.getExerciseCount() == 0) {
            exerciseDao.insertExercises(ExerciseLibrary.getDefaultExercises())
        }
    }

    // Workouts
    fun getAllWorkouts(): Flow<List<GymWorkout>> = gymWorkoutDao.getAllWorkoutsFlow()
    
    suspend fun getAllWorkoutsOnce(): List<GymWorkout> = gymWorkoutDao.getAllWorkouts()
    
    fun getRecentWorkouts(limit: Int = 10): Flow<List<GymWorkout>> = 
        gymWorkoutDao.getRecentWorkoutsFlow(limit)
    
    fun getWorkoutById(id: Long): Flow<GymWorkout?> = gymWorkoutDao.getWorkoutByIdFlow(id)
    
    suspend fun getWorkoutByIdOnce(id: Long): GymWorkout? = gymWorkoutDao.getWorkoutById(id)
    
    suspend fun insertWorkout(workout: GymWorkout): Long = gymWorkoutDao.insertWorkout(workout)
    
    suspend fun updateWorkout(workout: GymWorkout) = gymWorkoutDao.updateWorkout(workout)
    
    suspend fun deleteWorkout(workout: GymWorkout) = gymWorkoutDao.deleteWorkout(workout)
    
    suspend fun getActiveWorkout(): GymWorkout? = gymWorkoutDao.getActiveWorkout()
    
    fun getWorkoutsInRange(startTime: Long, endTime: Long): Flow<List<GymWorkout>> =
        gymWorkoutDao.getWorkoutsInRangeFlow(startTime, endTime)

    // Templates
    fun getAllTemplates(): Flow<List<WorkoutTemplate>> = workoutTemplateDao.getAllTemplatesFlow()
    
    fun getDefaultTemplates(): Flow<List<WorkoutTemplate>> = workoutTemplateDao.getDefaultTemplatesFlow()
    
    fun getCustomTemplates(): Flow<List<WorkoutTemplate>> = workoutTemplateDao.getCustomTemplatesFlow()
    
    suspend fun getTemplateById(id: Long): WorkoutTemplate? = workoutTemplateDao.getTemplateById(id)
    
    fun getTemplateByIdFlow(id: Long): Flow<WorkoutTemplate?> = workoutTemplateDao.getTemplateByIdFlow(id)
    
    suspend fun insertTemplate(template: WorkoutTemplate): Long = workoutTemplateDao.insertTemplate(template)
    
    suspend fun updateTemplate(template: WorkoutTemplate) = workoutTemplateDao.updateTemplate(template)
    
    suspend fun deleteTemplate(template: WorkoutTemplate) = workoutTemplateDao.deleteTemplate(template)
    
    suspend fun incrementTemplateUsage(templateId: Long) = workoutTemplateDao.incrementUsage(templateId)
    
    suspend fun initializeDefaultTemplates() {
        if (workoutTemplateDao.getTemplateCount() == 0) {
            val templates = DefaultTemplates.getPushPullLegs() + 
                           DefaultTemplates.getUpperLower() + 
                           listOf(DefaultTemplates.getFullBody())
            workoutTemplateDao.insertTemplates(templates)
        }
    }

    // Exercise History
    fun getHistoryForExercise(exerciseId: Long): Flow<List<ExerciseHistory>> = 
        exerciseHistoryDao.getHistoryForExerciseFlow(exerciseId)
    
    suspend fun getRecentHistoryForExercise(exerciseId: Long, limit: Int = 10): List<ExerciseHistory> =
        exerciseHistoryDao.getRecentHistoryForExercise(exerciseId, limit)
    
    suspend fun getBestOneRepMax(exerciseId: Long): ExerciseHistory? =
        exerciseHistoryDao.getBestOneRepMaxForExercise(exerciseId)
    
    suspend fun getBestWeight(exerciseId: Long): ExerciseHistory? =
        exerciseHistoryDao.getBestWeightForExercise(exerciseId)
    
    fun getPersonalRecords(): Flow<List<ExerciseHistory>> = exerciseHistoryDao.getPersonalRecordsFlow()
    
    suspend fun getPersonalRecordsSnapshot(): List<ExerciseHistory> = exerciseHistoryDao.getPersonalRecords()
    
    suspend fun recordExerciseHistory(
        exerciseId: Long,
        workoutId: Long,
        sets: List<WorkoutSet>
    ): PBResult? {
        val completedSets = sets.filter { it.isCompleted }
        if (completedSets.isEmpty()) return null
        
        val bestSet = completedSets.maxByOrNull { it.weight * it.reps }!!
        val totalVolume = completedSets.sumOf { it.weight * it.reps }
        val totalReps = completedSets.sumOf { it.reps }
        val estimatedOneRepMax = OneRepMaxCalculator.calculate(bestSet.weight, bestSet.reps)
        
        val previousBest = exerciseHistoryDao.getBestOneRepMaxForExercise(exerciseId)
        val isPersonalRecord = previousBest == null || estimatedOneRepMax > previousBest.estimatedOneRepMax
        
        val history = ExerciseHistory(
            exerciseId = exerciseId,
            workoutId = workoutId,
            date = System.currentTimeMillis(),
            bestWeight = bestSet.weight,
            bestReps = bestSet.reps,
            totalVolume = totalVolume,
            totalSets = completedSets.size,
            totalReps = totalReps,
            estimatedOneRepMax = estimatedOneRepMax,
            isPersonalRecord = isPersonalRecord
        )
        
        exerciseHistoryDao.insertHistory(history)
        
        return if (isPersonalRecord) {
            PBResult(
                exerciseId = exerciseId,
                weight = bestSet.weight,
                reps = bestSet.reps,
                estimatedOneRepMax = estimatedOneRepMax
            )
        } else null
    }
    
    data class PBResult(
        val exerciseId: Long,
        val weight: Double,
        val reps: Int,
        val estimatedOneRepMax: Double
    )

    // Progression suggestions
    suspend fun getProgressionSuggestion(
        exerciseId: Long,
        exerciseName: String,
        targetReps: IntRange = 8..12
    ): ProgressionSuggestion {
        val recentHistory = exerciseHistoryDao.getRecentHistoryForExercise(exerciseId, 5)
        return ProgressionEngine.suggestProgression(
            exerciseId = exerciseId,
            exerciseName = exerciseName,
            recentHistory = recentHistory,
            targetReps = targetReps
        )
    }

    // Weekly stats
    suspend fun getWeeklyGymStats(): GymWeeklyStats {
        val calendar = Calendar.getInstance().apply {
            set(Calendar.DAY_OF_WEEK, Calendar.MONDAY)
            set(Calendar.HOUR_OF_DAY, 0)
            set(Calendar.MINUTE, 0)
            set(Calendar.SECOND, 0)
            set(Calendar.MILLISECOND, 0)
        }
        val weekStart = calendar.timeInMillis
        val weekEnd = weekStart + 7 * 24 * 60 * 60 * 1000L
        
        val totalVolume = gymWorkoutDao.getTotalVolumeInRange(weekStart, weekEnd) ?: 0.0
        val workoutCount = gymWorkoutDao.getWorkoutCountInRange(weekStart, weekEnd)
        
        return GymWeeklyStats(
            weekStartDate = weekStart,
            totalVolume = totalVolume,
            workoutCount = workoutCount
        )
    }
}

data class GymWeeklyStats(
    val weekStartDate: Long,
    val totalVolume: Double,
    val workoutCount: Int
) {
    val totalVolumeFormatted: String
        get() = if (totalVolume >= 1000) {
            String.format("%.1fk kg", totalVolume / 1000)
        } else {
            String.format("%.0f kg", totalVolume)
        }
}

File: iosApp/GoSteady/Sources/Repositories/GymRepository.swift
import Foundation
import SwiftData

final class GymRepository {
    private let context: ModelContext

    init(context: ModelContext) {
        self.context = context
    }

    // MARK: - Workouts

    func fetchAllWorkouts(sortedBy order: SortOrder = .reverse) -> [GymWorkout] {
        let descriptor = FetchDescriptor<GymWorkout>(sortBy: [SortDescriptor(\.startTime, order: order)])
        return (try? context.fetch(descriptor)) ?? []
    }

    func fetchRecentWorkouts(limit: Int = 10) -> [GymWorkout] {
        var descriptor = FetchDescriptor<GymWorkout>(sortBy: [SortDescriptor(\.startTime, order: .reverse)])
        descriptor.fetchLimit = limit
        return (try? context.fetch(descriptor)) ?? []
    }

    func fetchWorkoutById(_ id: UUID) -> GymWorkout? {
        let descriptor = FetchDescriptor<GymWorkout>(predicate: #Predicate { $0.id == id })
        return try? context.fetch(descriptor).first
    }

    func fetchActiveWorkout() -> GymWorkout? {
        let descriptor = FetchDescriptor<GymWorkout>(predicate: #Predicate { $0.isCompleted == false })
        return try? context.fetch(descriptor).first
    }

    func fetchWorkoutsInRange(start: Date, end: Date) -> [GymWorkout] {
        let descriptor = FetchDescriptor<GymWorkout>(
            predicate: #Predicate { $0.startTime >= start && $0.startTime <= end },
            sortBy: [SortDescriptor(\.startTime, order: .reverse)]
        )
        return (try? context.fetch(descriptor)) ?? []
    }

    func fetchCompletedWorkoutsInRange(start: Date, end: Date) -> [GymWorkout] {
        let descriptor = FetchDescriptor<GymWorkout>(
            predicate: #Predicate { $0.startTime >= start && $0.startTime <= end && $0.isCompleted == true },
            sortBy: [SortDescriptor(\.startTime, order: .reverse)]
        )
        return (try? context.fetch(descriptor)) ?? []
    }

    func insertWorkout(_ workout: GymWorkout) {
        context.insert(workout)
        try? context.save()
    }

    func updateWorkout(_ workout: GymWorkout) {
        try? context.save()
    }

    func deleteWorkout(_ workout: GymWorkout) {
        context.delete(workout)
        try? context.save()
    }

    // MARK: - Templates

    func fetchAllTemplates() -> [WorkoutTemplate] {
        let descriptor = FetchDescriptor<WorkoutTemplate>(sortBy: [SortDescriptor(\.name)])
        return (try? context.fetch(descriptor)) ?? []
    }

    func fetchDefaultTemplates() -> [WorkoutTemplate] {
        let descriptor = FetchDescriptor<WorkoutTemplate>(
            predicate: #Predicate { $0.isDefault == true },
            sortBy: [SortDescriptor(\.name)]
        )
        return (try? context.fetch(descriptor)) ?? []
    }

    func fetchCustomTemplates() -> [WorkoutTemplate] {
        let descriptor = FetchDescriptor<WorkoutTemplate>(
            predicate: #Predicate { $0.isDefault == false },
            sortBy: [SortDescriptor(\.name)]
        )
        return (try? context.fetch(descriptor)) ?? []
    }

    func fetchTemplateById(_ id: UUID) -> WorkoutTemplate? {
        let descriptor = FetchDescriptor<WorkoutTemplate>(predicate: #Predicate { $0.id == id })
        return try? context.fetch(descriptor).first
    }

    func insertTemplate(_ template: WorkoutTemplate) {
        context.insert(template)
        try? context.save()
    }

    func updateTemplate(_ template: WorkoutTemplate) {
        try? context.save()
    }

    func deleteTemplate(_ template: WorkoutTemplate) {
        context.delete(template)
        try? context.save()
    }

    func incrementTemplateUsage(_ template: WorkoutTemplate) {
        template.timesUsed += 1
        template.lastUsed = Date()
        try? context.save()
    }

    func templateCount() -> Int {
        let descriptor = FetchDescriptor<WorkoutTemplate>()
        return (try? context.fetchCount(descriptor)) ?? 0
    }

    // MARK: - Exercise History

    func fetchHistoryForExercise(_ exerciseId: UUID) -> [ExerciseHistory] {
        let descriptor = FetchDescriptor<ExerciseHistory>(
            predicate: #Predicate { $0.exerciseId == exerciseId },
            sortBy: [SortDescriptor(\.date, order: .reverse)]
        )
        return (try? context.fetch(descriptor)) ?? []
    }

    func fetchRecentHistoryForExercise(_ exerciseId: UUID, limit: Int = 10) -> [ExerciseHistory] {
        var descriptor = FetchDescriptor<ExerciseHistory>(
            predicate: #Predicate { $0.exerciseId == exerciseId },
            sortBy: [SortDescriptor(\.date, order: .reverse)]
        )
        descriptor.fetchLimit = limit
        return (try? context.fetch(descriptor)) ?? []
    }

    func fetchBestOneRepMax(exerciseId: UUID) -> ExerciseHistory? {
        let descriptor = FetchDescriptor<ExerciseHistory>(
            predicate: #Predicate { $0.exerciseId == exerciseId },
            sortBy: [SortDescriptor(\.estimatedOneRepMax, order: .reverse)]
        )
        return try? context.fetch(descriptor).first
    }

    func fetchBestWeight(exerciseId: UUID) -> ExerciseHistory? {
        let descriptor = FetchDescriptor<ExerciseHistory>(
            predicate: #Predicate { $0.exerciseId == exerciseId },
            sortBy: [SortDescriptor(\.bestWeight, order: .reverse)]
        )
        return try? context.fetch(descriptor).first
    }

    func fetchPersonalRecords() -> [ExerciseHistory] {
        let descriptor = FetchDescriptor<ExerciseHistory>(
            predicate: #Predicate { $0.isPersonalRecord == true },
            sortBy: [SortDescriptor(\.date, order: .reverse)]
        )
        return (try? context.fetch(descriptor)) ?? []
    }

    func insertHistory(_ history: ExerciseHistory) {
        context.insert(history)
        try? context.save()
    }

    func updateHistory(_ history: ExerciseHistory) {
        try? context.save()
    }

    /// Record exercise history from completed sets and detect personal records.
    func recordExerciseHistory(
        exerciseId: UUID,
        workoutId: UUID,
        sets: [WorkoutSet]
    ) -> GymPBResult? {
        let completed = sets.filter(\.isCompleted)
        guard !completed.isEmpty else { return nil }

        let bestSet = completed.max(by: { ($0.weight * Double($0.reps)) < ($1.weight * Double($1.reps)) })!
        let totalVolume = completed.reduce(0.0) { $0 + $1.weight * Double($1.reps) }
        let totalReps = completed.reduce(0) { $0 + $1.reps }
        let estimated1RM = OneRepMaxCalculator.calculate(weight: bestSet.weight, reps: bestSet.reps)

        let previousBest = fetchBestOneRepMax(exerciseId: exerciseId)
        let isPR = previousBest == nil || estimated1RM > (previousBest?.estimatedOneRepMax ?? 0)

        let history = ExerciseHistory(
            exerciseId: exerciseId,
            workoutId: workoutId,
            date: Date(),
            bestWeight: bestSet.weight,
            bestReps: bestSet.reps,
            totalVolume: totalVolume,
            totalSets: completed.count,
            totalReps: totalReps,
            estimatedOneRepMax: estimated1RM,
            isPersonalRecord: isPR
        )
        insertHistory(history)

        return isPR ? GymPBResult(
            exerciseId: exerciseId,
            weight: bestSet.weight,
            reps: bestSet.reps,
            estimatedOneRepMax: estimated1RM
        ) : nil
    }

    // MARK: - Progression

    func progressionSuggestion(
        exerciseId: UUID,
        exerciseName: String,
        targetReps: ClosedRange<Int> = 8...12
    ) -> ProgressionSuggestion {
        let history = fetchRecentHistoryForExercise(exerciseId, limit: 5)
        return ProgressionEngine.suggestProgression(
            exerciseId: exerciseId,
            exerciseName: exerciseName,
            recentHistory: history,
            targetReps: targetReps
        )
    }

    // MARK: - Weekly Stats

    func weeklyGymStats() -> GymWeeklyStats {
        let range = Date.thisWeekRange
        let workouts = fetchCompletedWorkoutsInRange(start: range.start, end: range.end)
        let totalVolume = workouts.reduce(0.0) { $0 + $1.totalVolume }
        return GymWeeklyStats(
            weekStartDate: range.start,
            totalVolume: totalVolume,
            workoutCount: workouts.count
        )
    }
}

struct GymPBResult {
    let exerciseId: UUID
    let weight: Double
    let reps: Int
    let estimatedOneRepMax: Double
}

struct GymWeeklyStats {
    let weekStartDate: Date
    let totalVolume: Double
    let workoutCount: Int

    var totalVolumeFormatted: String {
        FormatUtils.formatVolume(totalVolume)
    }
}
